src/runtime/buffer.h

```diff
@@ -107,7 +107,14 @@ static inline boolean buffer_extend(buffer b, bytes len)
     // xxx - pad to pagesize
     if (b->length < (b->end + len)) {
         bytes new_len = 2 * (b->end - b->start + len);
-        return (buffer_set_capacity(b, new_len) == new_len);
+        if (new_len > b->length) {
+            return (buffer_set_capacity(b, new_len) == new_len);
+        } else {
+            /* no need to resize, move current contents to the beginning of the allocated memory */
+            runtime_memcpy(b->contents, b->contents + b->start, b->end - b->start);
```

Comment:
I think this is fine but just thinking this through: the source and destination ranges may overlap, but it should be no issue as data is being moved from a higher to lower location, right?